### PR TITLE
Fix and improve REALM fine-tuning

### PR DESCRIPTION
**Pull request by qqaatw**
# What does this PR do?

This PR 
1. adds `block_embedding_to` function to `RealmForOpenQA`, which allows users to send `block_emb` variable to a specific device. 
3. adds an input argument `block_mask` for reader that replaces previous used `token_type_ids` as the candidate mask. As a result, the reader logits at the second `[SEP]` position of the sequence will be masked out, which would prevent `[SEP]` from occurring on the predicted answer spans and matches the original TF implementation.

Their difference:
```
sentence: "[CLS] Hello How are you? [SEP] Good Job [SEP]"
token_type_ids = [0, 0, 0, 0, 0, 0, 0, 1, 1, 1] # the last 1 is the second [SEP] token
block_mask     = [0, 0, 0, 0, 0, 0, 0, 1, 1, 0]
```

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

File: src/transformers/models/realm/modeling_realm.py [hunk 1]
```diff
@@ -1736,6 +1736,41 @@ def __init__(self, config, retriever=None):
 
         self.post_init()
 
+    def to(self, *args, **kwargs):
+        """Override `torch.module.to` in order to prevent `self.block_emb`, which would largely consume gpu resources, from
```
Comment (qqaatw) [inline src/transformers/models/realm/modeling_realm.py:1740]:
```suggestion
        """Override `torch.nn.Module.to` in order to prevent `self.block_emb`, which would largely consume gpu resources, from
```

File: src/transformers/models/realm/modeling_realm.py [hunk 2]
```diff
@@ -1736,6 +1736,41 @@ def __init__(self, config, retriever=None):
 
         self.post_init()
 
+    def to(self, *args, **kwargs):
+        """Override `torch.nn.Module.to` in order to prevent `self.block_emb`, which would largely consume gpu resources, from
```
Comment (patrickvonplaten) [inline src/transformers/models/realm/modeling_realm.py:1740]: could you add a link to the docstring of the original method here as well?

**Review by patrickvonplaten — COMMENTED** [review]
In general, I'm fine with the change @qqaatw - thank you! :-) 

Do you know by how much this change slows down the inference time of the model?

**Comment by qqaatw** [comment]
@patrickvonplaten - The inference time of NQ evaluation dataset is `8:25` when `block_emb` is on GPU, while the inference time is `21:07` when `block_emb` is on CPU, which is roughly 3 times slower than on GPU. The test was performed on single RTX 2080Ti GPU.

Do you think we should provide an option to let users decide whether to send `block_emb` to GPU? Considering that much slower inference speed...

**Comment by patrickvonplaten** [comment]
> evaluation

That is indeed a pretty big slow down...given such inference times I think we should have `block_emb` by default on the same device than the other parameters (so we should not overwrite the `.to(...)` method I think). I would be happy with adding a `def block_embedding_to(device)` method that lets someone easily move only `block_emb` to CPU/GPU - what do you think?

**Comment by qqaatw** [comment]
Considering most users would expect the model can be fine-tuned on single gpu with 12GiB memory as the paper states, I'm a bit worried that if we by default have `block_emb` on the same device (let's say a 2080Ti) as other parameters, when users call `block_embedding_to` to send it back to CPU, the cuda memory occupied by `block_emb` might not be able to be released as expected. 

To my knowledge, `torch.Tensor.to` will have the original tensor untouched and create a new tensor on the new device (so I think it's more like a copy rather than a transfer), and the way I know to release the memory occupied by the original tensor is delete the tensor until it is cleared by Python GC or by manually calling `torch.cuda.empty_cache()`. However, because I'm not very familiar with how PyTorch cuda context works, and I've seen in some circumstances that even though the tensor object was deleted and manually cleared, occupied spaces are still not be freed up, I don't have 100% confidence that the model can be trained on the GPU with 12GiB memory normally.

So to sum up, I think we should still override `model.to` and add a `def block_embedding_to(device)` as you suggest, this way users can choose whether they want to move `block_emb` to the GPU if they have one with big enough memory.

Does this make sense to you?

File: src/transformers/models/realm/modeling_realm.py [hunk 3]
```diff
@@ -875,8 +875,7 @@ def mask_to_score(mask):
         hidden_states = self.dense_intermediate(hidden_states)
         # [reader_beam_size, max_sequence_len, span_hidden_size]
         start_projection, end_projection = hidden_states.chunk(2, dim=-1)
-        block_mask = token_type_ids.detach().clone()
-        block_mask[:, -1] = 0
+
         candidate_starts, candidate_ends, candidate_mask = span_candidates(block_mask)
```
Comment (qqaatw) [inline src/transformers/models/realm/modeling_realm.py:879]: We previously used `token_type_ids` as the `block_mask`, and now the `block_mask` is computed in the retriever.

Comment (qqaatw) [inline src/transformers/models/realm/modeling_realm.py:879]: We previously used `type_token_ids` as `block_mask`, and now the `block_mask` is computed in the retriever.

**Comment by qqaatw** [comment]
@patrickvonplaten - I noticed that the `block_records.npy` file has not been uploaded to model repos having an `openqa` suffix. Could you help with this so that people are able to fine-tune the model. Thanks.

File: tests/test_modeling_realm.py [hunk 4]
```diff
@@ -409,6 +409,15 @@ def test_training(self):
...
         loss.backward()
 
+        self.assertEqual(model.block_emb.device, torch.device("cpu"))
+
+        device = torch.device(torch_device)
+        with torch.no_grad():
+            model.block_embedding_to(device)
+            model(**inputs)
+
+        self.assertEqual(model.block_emb.device.type, device.type)
```
Comment (qqaatw) [inline tests/test_modeling_realm.py:419]: Compare `device.type` rather than `device` as `torch_device` doesn't have a device index while `model.block_emb.device` has.

**Comment by patrickvonplaten** [comment]
> 

Hey @qqaatw,

Is it ok if we first move the repos to the official Google org and then add the indices? This should be a bit easier :-)

File: src/transformers/models/realm/modeling_realm.py [hunk 5]
```diff
@@ -1543,6 +1542,7 @@ def forward(
         head_mask=None,
         inputs_embeds=None,
         relevance_score=None,
+        block_mask=None,
```
Comment (patrickvonplaten) [inline src/transformers/models/realm/modeling_realm.py:1545]: It seems like `token_type_ids` is not used here anymore at all. This is quite backwards compatibility breaking. Is it intended? Should `block_mask` replace `token_type_ids`?

File: src/transformers/models/realm/modeling_realm.py [hunk 6]
```diff
@@ -1737,11 +1742,59 @@ def __init__(self, config, retriever=None):
...
+        """Send `self.block_emb` to a specific device. This is useful for accelerating training and inference speed.
+
+        Args:
+            device (`str` or `torch.device`):
+                The device that `self.block_emb` will be sent to.
+        """
+
+        self.block_emb = self.block_emb.to(device)
+
+    def to(self, *args, **kwargs):
```
Comment (patrickvonplaten) [inline src/transformers/models/realm/modeling_realm.py:1760]: I think we should allow the `block_emb` to be on CUDA. If we overwrite `to(...)` here this means that one cannot the full forward pass on GPU. Given that the computation is much faster on GPU for `block_emb`, I don't think we should add such a method here. IMO, it'd be enough to have `block_embedding_to(...)` no?

File: src/transformers/models/realm/retrieval_realm.py [hunk 7]
```diff
@@ -115,7 +117,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
             )
         block_records = np.load(block_records_path, allow_pickle=True)
 
-        tokenizer = RealmTokenizer.from_pretrained(pretrained_model_name_or_path, *init_inputs, **kwargs)
```
Comment (patrickvonplaten) [inline src/transformers/models/realm/retrieval_realm.py:118]: Let's maybe use `AutoTokenizer` here so that the user can decide between Fast and Non-Fast

File: src/transformers/models/realm/modeling_realm.py (hunk 5 above)
Comment (qqaatw) [reply src/transformers/models/realm/modeling_realm.py:1545]: `token_type_ids` is still used in the realm(bert) model inside `RealmReader`. We just don't use `token_type_ids` as the candidate mask for convenience and use the newly added `block_mask` instead, because I realized the former approach would result in `[SEP]` token occurring in the predicted answer spans, which is explained [here](https://github.com/huggingface/transformers/pull/15297#issue-1111824123).

This is indeed a small breaking change. However, since users would not directly leverage `RealmReader` (`RealmReader` 
is an inner model of `RealmForOpenQA`, and the `relevance_score` input of `RealmReader` needs to be firstly computed in `RealmForOpenQA`), IMO I think this would not affect existing codes.

File: src/transformers/models/realm/modeling_realm.py (hunk 6 above)
Comment (qqaatw) [reply src/transformers/models/realm/modeling_realm.py:1760]: Okay. I tested if we firstly call model.to("cuda") and then call `model.block_embedding_to("cpu")`, the gpu memory occupied by `block_emb` could be released, and the training ran normally on single RTX 2080Ti.

Note that one potential scenario is that if one's gpu has only 6 GiB memory like RTX 2060, the model will not be able to run training/inference on GPU following the above workflow because `model.to("cuda")` will consume roughly 8~9 GiB memory if `block_emb` is also sent to GPU, and the cuda out of memory error will occur.

Instead, if we override `model.to(...)`, we can do the following workflows:

with 40 GiB gpu memory like A100:
```
# training / inference (I think this will do full forward pass on GPU? let me know if I'm wrong)
model.to("cuda")
model.block_embedding_to("cuda")
```

with 11 GiB gpu memory like RTX 2080Ti:
```
# training (memory will not enough to store gradients)
model.to("cuda")

# inference
model.to("cuda")
model.block_embedding_to("cuda")
```

with 6 GiB gpu memory like RTX 2060:

```
# inference (not sure if training is possible)
model.to("cuda") # roughly consume 4 GiB
```

We can probably add FAISS support in the future, which is more compatible with PyTorch than ScaNN.

**Comment by qqaatw** [comment]
@patrickvonplaten Sure, but can I keep these checkpoints also on my namespace for testing (I can set private for them)? I will not be able to update checkpoints within Google org namespace right?

**Comment by qqaatw** [comment]
Hi @patrickvonplaten, do the above comments make sense to you? or I can just follow your guide. Sorry for the late update.

File: src/transformers/models/realm/modeling_realm.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/transformers/models/realm/modeling_realm.py:1760]: Hey @qqaatw, 

I still don't think it's worth trying to optimize for a potential 6GB GPU. Most GPU's people work with have at least 16GB nowadays IMO. Adding a specific `def to(...)` method is very much not in line with other models in the library and might be pretty confusing to the user (note that 99% of users don't read code). So I would prefer to leave as it is here now :-)

Comment (qqaatw) [reply src/transformers/models/realm/modeling_realm.py]: OK that makes sense. I've remove it :)

File: src/transformers/models/realm/modeling_realm.py [hunk 8]
```diff
@@ -1737,11 +1742,21 @@ def __init__(self, config, retriever=None):
         self.post_init()
 
     @property
-    def beam_size(self):
+    def searcher_beam_size(self):
```
Comment (patrickvonplaten) [inline src/transformers/models/realm/modeling_realm.py:1745]: This breaks backward compatibility here - `realm.beam_size` would now error out. Can't we just leave the name?

File: src/transformers/models/realm/modeling_realm.py [hunk 9]
```diff
@@ -1737,11 +1742,21 @@ def __init__(self, config, retriever=None):
...
         if self.training:
             return self.config.searcher_beam_size
         return self.config.reader_beam_size
 
+    def block_embedding_to(self, device):
```
Comment (patrickvonplaten) [inline src/transformers/models/realm/modeling_realm.py:1750]: Ok for me even though I don't think it's really necessary

File: src/transformers/models/realm/modeling_realm.py (hunk 8 above)
Comment (patrickvonplaten) [reply src/transformers/models/realm/modeling_realm.py:1745]: That's the only thing which would be nice to revert - the rest looks good to me!

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/models/realm/modeling_realm.py (hunk 9 above)
Comment (qqaatw) [reply src/transformers/models/realm/modeling_realm.py:1750]: @patrickvonplaten - Do you think if it is better that I just add some tips in the model docs, e.g. telling users they can manually send `block_emb` back to cpu if their gpu memory is not sufficient, and remove this function.

We still need to prompt users how they can fine-tune the model on a 12Gb memory GPU as promised by the paper.

Comment (patrickvonplaten) [reply src/transformers/models/realm/modeling_realm.py:1750]: I think the best would be to add a comment under `Implementation details` on the model doc above this section: https://github.com/huggingface/transformers/blob/master/docs/source/model_doc/realm.mdx#realmconfig 

But it's also ok for me to leave as is now - your choice :-)

**Review by patil-suraj — APPROVED** [review]
LGTM! Thank you for adding this.
